service/raft/group0_state_id_handler.cc

```diff
@@ -6,15 +6,124 @@
  * SPDX-License-Identifier: AGPL-3.0-or-later
  */
 #include "service/raft/group0_state_id_handler.hh"
+#include "db/config.hh"
+#include "service/raft/raft_address_map.hh"
+#include "compaction/compaction_manager.hh"
 #include "gms/gossiper.hh"
 #include "log.hh"
+#include "raft/server.hh"
+#include "replica/database.hh"
+#include "service/raft/raft_group_registry.hh"
 
 namespace service {
 
 static logging::logger slogger("group0_tombstone_gc_handler");
 
-group0_state_id_handler::group0_state_id_handler(gms::gossiper& gossiper)
-    : _gossiper(gossiper) {
+raft::server* group0_server_accessor::get_server() const {
+    return _raft_gr.find_server(_group0_id);
+}
+
+lowres_clock::duration group0_state_id_handler::get_broadcast_interval(const replica::database& db) {
+    return std::chrono::milliseconds{db.get_config().group0_tombstone_gc_interval_in_ms()};
+}
+
+group0_state_id_handler::group0_state_id_handler(
+        replica::database& local_db, gms::gossiper& gossiper, const raft_address_map& address_map, group0_server_accessor server_accessor)
+    : _local_db(local_db)
+    , _gossiper(gossiper)
+    , _address_map(address_map)
+    , _server_accessor(server_accessor)
+    , _check_interval(get_broadcast_interval(local_db)) {
+}
+
+group0_state_id_handler::~group0_state_id_handler() {
+    if (!_stopped) {
+        on_fatal_internal_error(slogger, "group0_state_id_handler is being destroyed without being stopped");
+    }
+}
+
+void group0_state_id_handler::start() {
+    if (this_shard_id() != 0) {
+        on_fatal_internal_error(slogger, "group0_state_id_handler must be started on shard 0");
+    }
+
+    if (_stopped) {
+        return;
+    }
+
+    _timer.set_callback([this] {
+        std::invoke([this] {
```

Comment:
it's because of the control flow, to be able to use returns - the thing is that I need to re-set the timer every time (even if the check "failed") so it is a bit tricky otherwise (when the code was async before, I used `.finally()` for that, but now the code doesn't actually need to be async so the invoke is just to allow direct return inside the lambda) ... but I guess I could use a private / static helper function instead

Reply:
Ah, I see. makes sense

Reply:
or I can actually use `defer` for the timer rest, that should allow to do this every time even when using the returns

Reply:
Such huge lambdas (I'm referring to the outer one) are best split to a named function. Stack traces from the timer will show group0_state_handler::start()::lambda#1::operator() or something, and someone will have to understand that it's really the timer.

Reply:
Right, moved the long lambda contents to a separate method